### PR TITLE
Fix anonymous downloads URL generation

### LINKED ISSUE #6778: Anonymous downloads now working via web UI on bucket with read-only policy
This appears to be a regression, similar to the one in #5076.

We are having this issue when running release 2018-11-06T01:01:02Z, this should already be fixed there, correct? I can see that we are not generating the token, as the URL ends with `?token=%27%27`. Prior to upgrading, this same URL did work.

If I change the URL so that instead of `https://host/minio/download/…?token=%27%27` it is `https://host/…`, then the file can be viewed/downloaded. Just removing the `?token=%27%27` also does not work for the download, that just takes me to the login screen.

## Expected Behavior
Clicking download should on a file that I can see should download the file.

## Current Behavior
Clicking download should on a file that I can see results in the error message "Authentication failed, check your access credentials".

## Possible Solution
A work around is presented above, use the direct URL instead of the download URL. This is obviously inconvenient for users.

## Steps to Reproduce (for bugs)
* Create bucket using web ui
* Upload file into bucket using web ui
* Create * read-only policy using web ui
* Copy URL from browser and open in another incognito window
* Attempt to download one of the files using web ui

## Context
Our users are no longer able to download files.

## Regression
Is this issue a regression? *Yes*

This was working before the upgrade, we were running the 2018-10-25 release.

## Your Environment
* Version used (`minio version`): 2018-11-06T01:01:02Z

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
Fix anonymous downloads URL generation

## Motivation and Context
Fixes #6778

## Regression
Yes

## How Has This Been Tested?
Follow the steps mentioned in issue #6778 by building the PR locally. 

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :x: [more...](#a306f464-9523-41a5-9257-e84df1411a3b) |
<a name="a306f464-9523-41a5-9257-e84df1411a3b"></a>
### 6800-4da7683/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:31797
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp e987a68f8b9e:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 1 minutes and 44 seconds
(2/13) Running aws-sdk-go tests ... done in 1 seconds
(3/13) Running aws-sdk-java tests ... done in 3 seconds
(4/13) Running aws-sdk-php tests ... done in 44 seconds
(5/13) Running aws-sdk-ruby tests ... done in 11 seconds
(6/13) Running mc tests ... done in 1 minutes and 16 seconds
(7/13) Running minio-dotnet tests ... FAILED in 5 minutes and 25 seconds

Unhandled Exception: System.AggregateException: One or more errors occurred. (One or more errors occurred. (Minio API responded with message=We encountered an internal error, please try again.)) ---> System.AggregateException: One or more errors occurred. (Minio API responded with message=We encountered an internal error, please try again.) ---> Minio.Exceptions.MinioException: Minio API responded with message=We encountered an internal error, please try again.
   at Minio.MinioClient.ParseError(IRestResponse response) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 363
   at Minio.MinioClient.HandleIfErrorResponse(IRestResponse response, IEnumerable`1 handlers, DateTime startTime) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 505
   at Minio.MinioClient.ExecuteTaskAsync(IEnumerable`1 errorHandlers, IRestRequest request, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 351
   at Minio.MinioClient.PutObjectAsync(String bucketName, String objectName, String uploadId, Int32 partNumber, Byte[] data, Dictionary`2 metaData, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/ApiEndpoints/ObjectOperations.cs:line 483
   at Minio.MinioClient.PutObjectAsync(String bucketName, String objectName, Stream data, Int64 size, String contentType, Dictionary`2 metaData, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/ApiEndpoints/ObjectOperations.cs:line 229
   at Minio.Functional.Tests.FunctionalTest.PutObject_Task(MinioClient minio, String bucketName, String objectName, String fileName, String contentType, Int64 size, Dictionary`2 metaData, MemoryStream mstream) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 740
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at Minio.Functional.Tests.FunctionalTest.RemoveObjects_Test2(MinioClient minio) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 1738
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at Minio.Functional.Tests.FunctionalTest.Main(String[] args) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 218

Executed 6 out of 13 tests successfully.
```

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#43a36148-2e92-44c7-808b-def7d0ba338e) |
<a name="43a36148-2e92-44c7-808b-def7d0ba338e"></a>
### 6800-ee85152/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:31655
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 5b2c6feb0480:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 2 minutes and 3 seconds
(2/13) Running aws-sdk-go tests ... done in 1 seconds
(3/13) Running aws-sdk-java tests ... done in 2 seconds
(4/13) Running aws-sdk-php tests ... done in 50 seconds
(5/13) Running aws-sdk-ruby tests ... done in 46 seconds
(6/13) Running mc tests ... done in 1 minutes and 40 seconds
(7/13) Running minio-dotnet tests ... FAILED in 6 minutes and 46 seconds

Unhandled Exception: System.AggregateException: One or more errors occurred. (One or more errors occurred. (Minio API responded with message=Unsuccessful response from server without XML error: The operation has timed out.) (Minio API responded with message=Unsuccessful response from server without XML error: The operation has timed out.) (Minio API responded with message=Unsuccessful response from server without XML error: The operation has timed out.)) ---> System.AggregateException: One or more errors occurred. (Minio API responded with message=Unsuccessful response from server without XML error: The operation has timed out.) (Minio API responded with message=Unsuccessful response from server without XML error: The operation has timed out.) (Minio API responded with message=Unsuccessful response from server without XML error: The operation has timed out.) ---> Minio.Exceptions.InternalClientException: Minio API responded with message=Unsuccessful response from server without XML error: The operation has timed out.
   at Minio.MinioClient.ParseError(IRestResponse response) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 450
   at Minio.MinioClient.HandleIfErrorResponse(IRestResponse response, IEnumerable`1 handlers, DateTime startTime) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 505
   at Minio.MinioClient.ExecuteTaskAsync(IEnumerable`1 errorHandlers, IRestRequest request, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 351
   at Minio.MinioClient.PutObjectAsync(String bucketName, String objectName, String uploadId, Int32 partNumber, Byte[] data, Dictionary`2 metaData, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/ApiEndpoints/ObjectOperations.cs:line 483
   at Minio.MinioClient.PutObjectAsync(String bucketName, String objectName, Stream data, Int64 size, String contentType, Dictionary`2 metaData, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/ApiEndpoints/ObjectOperations.cs:line 229
   at Minio.Functional.Tests.FunctionalTest.PutObject_Task(MinioClient minio, String bucketName, String objectName, String fileName, String contentType, Int64 size, Dictionary`2 metaData, MemoryStream mstream) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 740
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at Minio.Functional.Tests.FunctionalTest.RemoveObjects_Test2(MinioClient minio) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 1738
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at Minio.Functional.Tests.FunctionalTest.Main(String[] args) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 218

Executed 6 out of 13 tests successfully.
```

**Review by poornas — DISMISSED** [review]
LGTM

File: cmd/web-handlers.go [hunk 1]
```diff
@@ -1452,8 +1452,30 @@ type PresignedGetRep struct {
...
+			return nil
+		}
 		return toJSONError(authErr)
 	}
+
+	if args.BucketName == "" || args.ObjectName == "" {
+		return &json2.Error{
+			Message: "Bucket and Object are mandatory arguments.",
+		}
+	}
```
Comment (krishnasrinivas) [inline cmd/web-handlers.go]: This is a regression bug in the browser code, i.e browser should not call presignedGet when the user is not logged-in. Fix is:
```
diff --git a/browser/app/js/objects/actions.js b/browser/app/js/objects/actions.js
index 3268ecde..f96002e6 100644
--- a/browser/app/js/objects/actions.js
+++ b/browser/app/js/objects/actions.js
@@ -279,7 +279,7 @@ export const downloadObject = object => {
     } else {
       const url = `${
         window.location.origin
-      }${minioBrowserPrefix}/download/${currentBucket}/${encObjectName}?token=''`
+      }${minioBrowserPrefix}/download/${currentBucket}/${encObjectName}?token=`
       window.location = url
     }
   }
```

Comment (harshavardhana) [reply cmd/web-handlers.go]: oh @krishnasrinivas

Comment (krishnasrinivas) [reply cmd/web-handlers.go]: do you want to fix the code for zip also:
```
--- a/browser/app/js/objects/actions.js
+++ b/browser/app/js/objects/actions.js
@@ -308,7 +308,7 @@ export const downloadCheckedObjects = () => {
       objects: getCheckedList(state)
     }
     if (!web.LoggedIn()) {
-      const requestUrl = location.origin + "/minio/zip?token=''"
+      const requestUrl = location.origin + "/minio/zip?token="
       downloadZip(requestUrl, req, dispatch)
     } else {
       return web
```

Comment (harshavardhana) [reply cmd/web-handlers.go]: yes fixing all of it @krishnasrinivas @poornas

File: browser/app/js/objects/actions.js [hunk 2]
```diff
@@ -60,10 +56,7 @@ export const appendList = (objects, marker, isTruncated) => ({
 
 export const fetchObjects = append => {
   return function(dispatch, getState) {
-    const {
-      buckets: { currentBucket },
-      objects: { currentPrefix, marker }
-    } = getState()
+    const {buckets: {currentBucket}, objects: {currentPrefix, marker}} = getState()
```
Comment (krishnasrinivas) [inline browser/app/js/objects/actions.js:59]: the previous deconstruction style was more readable

**Review by krishnasrinivas — DISMISSED** [review]

**Review by poornas — DISMISSED** [review]

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#998a3050-3015-4f9a-8bfa-b168a55a2519) |
<a name="998a3050-3015-4f9a-8bfa-b168a55a2519"></a>
### 6800-15f3a40/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:30056
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 42dda3f5e4c7:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 1 minutes and 59 seconds
(2/13) Running aws-sdk-go tests ... done in 3 seconds
(3/13) Running aws-sdk-java tests ... done in 3 seconds
(4/13) Running aws-sdk-php tests ... done in 57 seconds
(5/13) Running aws-sdk-ruby tests ... done in 47 seconds
(6/13) Running mc tests ... done in 1 minutes and 46 seconds
(7/13) Running minio-dotnet tests ... FAILED in 6 minutes and 23 seconds

Unhandled Exception: System.AggregateException: One or more errors occurred. (One or more errors occurred. (Minio API responded with message=Unsuccessful response from server without XML error: The operation has timed out.) (Minio API responded with message=Unsuccessful response from server without XML error: The operation has timed out.) (Minio API responded with message=Unsuccessful response from server without XML error: The operation has timed out.) (Minio API responded with message=Unsuccessful response from server without XML error: The operation has timed out.) (Minio API responded with message=Unsuccessful response from server without XML error: The operation has timed out.) (Minio API responded with message=Unsuccessful response from server without XML error: The operation has timed out.) (Minio API responded with message=Unsuccessful response from server without XML error: The operation has timed out.) (Minio API responded with message=Unsuccessful response from server without XML error: The operation has timed out.) (Minio API responded with message=Unsuccessful response from server without XML error: The operation has timed out.) (Minio API responded with message=Unsuccessful response from server without XML error: The operation has timed out.) (Minio API responded with message=Unsuccessful response from server without XML error: The operation has timed out.) (Minio API responded with message=Unsuccessful response from server without XML error: The operation has timed out.) (Minio API responded with message=Unsuccessful response from server without XML error: The operation has timed out.)) ---> System.AggregateException: One or more errors occurred. (Minio API responded with message=Unsuccessful response from server without XML error: The operation has timed out.) (Minio API responded with message=Unsuccessful response from server without XML error: The operation has timed out.) (Minio API responded with message=Unsuccessful response from server without XML error: The operation has timed out.) (Minio API responded with message=Unsuccessful response from server without XML error: The operation has timed out.) (Minio API responded with message=Unsuccessful response from server without XML error: The operation has timed out.) (Minio API responded with message=Unsuccessful response from server without XML error: The operation has timed out.) (Minio API responded with message=Unsuccessful response from server without XML error: The operation has timed out.) (Minio API responded with message=Unsuccessful response from server without XML error: The operation has timed out.) (Minio API responded with message=Unsuccessful response from server without XML error: The operation has timed out.) (Minio API responded with message=Unsuccessful response from server without XML error: The operation has timed out.) (Minio API responded with message=Unsuccessful response from server without XML error: The operation has timed out.) (Minio API responded with message=Unsuccessful response from server without XML error: The operation has timed out.) (Minio API responded with message=Unsuccessful response from server without XML error: The operation has timed out.) ---> Minio.Exceptions.InternalClientException: Minio API responded with message=Unsuccessful response from server without XML error: The operation has timed out.
   at Minio.MinioClient.ParseError(IRestResponse response) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 450
   at Minio.MinioClient.HandleIfErrorResponse(IRestResponse response, IEnumerable`1 handlers, DateTime startTime) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 505
   at Minio.MinioClient.ExecuteTaskAsync(IEnumerable`1 errorHandlers, IRestRequest request, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 351
   at Minio.MinioClient.PutObjectAsync(String bucketName, String objectName, String uploadId, Int32 partNumber, Byte[] data, Dictionary`2 metaData, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/ApiEndpoints/ObjectOperations.cs:line 483
   at Minio.MinioClient.PutObjectAsync(String bucketName, String objectName, Stream data, Int64 size, String contentType, Dictionary`2 metaData, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/ApiEndpoints/ObjectOperations.cs:line 229
   at Minio.Functional.Tests.FunctionalTest.PutObject_Task(MinioClient minio, String bucketName, String objectName, String fileName, String contentType, Int64 size, Dictionary`2 metaData, MemoryStream mstream) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 740
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at Minio.Functional.Tests.FunctionalTest.RemoveObjects_Test2(MinioClient minio) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 1738
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at Minio.Functional.Tests.FunctionalTest.Main(String[] args) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 218

Executed 6 out of 13 tests successfully.
```

File: browser/app/js/objects/actions.js (hunk 2 above)
Comment (harshavardhana) [reply browser/app/js/objects/actions.js:59]: I actually used esformatter it auto formatted

**Comment by harshavardhana** [comment]
Please review again @krishnasrinivas @poornas tests fixed.

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Review by poornas — APPROVED** [review]

**Review by krishnasrinivas — APPROVED** [review]

**Pull request merged**